File: SenLib/Sen3/FileFixes/r3000_dat.cs
using HyoutaUtils;
using HyoutaUtils.Streams;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace SenLib.Sen3.FileFixes {
	class r3000_dat : FileMod {
		public string GetDescription() {
			return "Terminology fixes in West Lamare Highway 2.";
		}

		public IEnumerable<FileModResult> TryApply(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x9bde3d287daabc7dul, 0xf76207d20b04da2aul, 0xd1c7120au));
			if (file == null) {
				return null;
			}

			var bin = file.CopyToMemory();
			var patcher = new SenScriptPatcher(bin);

			// port city -> Port City
			bin.Position = 0x40c3;
			bin.WriteUInt8(0x50);
			bin.Position = 0x40c8;
			bin.WriteUInt8(0x43);

			// The Juno Naval Fortress -> Juno Naval Fortress
			patcher.RemovePartialCommand(0xbd09, 0x94, 0xbd19, 0x4);

			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/r3000.dat", bin) };
		}

		public IEnumerable<FileModResult> TryRevert(FileStorage storage) {
			var file = storage.TryGetDuplicate(new HyoutaUtils.Checksum.SHA1(0x9bde3d287daabc7dul, 0xf76207d20b04da2aul, 0xd1c7120au));
			if (file == null) {
				return null;
			}
			return new FileModResult[] { new FileModResult("data/scripts/scena/dat_en/r3000.dat", file) };
		}
	}
}

File: native/sen3/FileFixes/r3000_dat.cpp
#include <filesystem>
#include <string_view>
#include <vector>

#include "file_getter.h"

#include "p3a/pack.h"
#include "p3a/structs.h"
#include "sen/sen_script_patcher.h"
#include "sha1.h"

namespace SenLib::Sen3::FileFixes::r3000_dat {
std::string_view GetDescription() {
    return "Terminology fixes in West Lamare Highway 2.";
}

bool TryApply(const SenPatcher::GetCheckedFileCallback& getCheckedFile,
              std::vector<SenPatcher::P3APackFile>& result) {
    try {
        auto file = getCheckedFile(
            "data/scripts/scena/dat_en/r3000.dat",
            50865,
            SenPatcher::SHA1FromHexString("9bde3d287daabc7df76207d20b04da2ad1c7120a"));
        if (!file) {
            return false;
        }

        auto& bin = file->Data;

        SenScriptPatcher patcher(bin);

        // port city -> Port City
        bin[0x40c3] = 0x50;
        bin[0x40c8] = 0x43;

        // The Juno Naval Fortress -> Juno Naval Fortress
        patcher.RemovePartialCommand(0xbd09, 0x94, 0xbd19, 0x4);

        result.emplace_back(std::move(bin), file->Filename, SenPatcher::P3ACompressionType::LZ4);

        return true;
    } catch (...) {
        return false;
    }
}
} // namespace SenLib::Sen3::FileFixes::r3000_dat
